File: packages/modules/blog/src/Models/Domain.fs
module Blog.Domain

// The blog module's data model — the reusable core (posts + comments + tags).
// Extracted from the microblog app; identity is NOT here (shared app-level layer),
// so ItemComment references it via the decoupled Hedge.Interface.IdentityRef.

open Hedge.Interface

// "Item" derives its table name (items -> blog_items when mounted); no [<Table>]
// override needed. Named neutrally: the app's name ("microblog") shouldn't live in
// the reusable module.
type Item = {
    Id: PrimaryKey<string>
    Title: string
    Link: Link option
    Image: Image option
    Extract: RichContent option
    OwnerComment: RichContent
    /// The article's own date (drives display, sort and day-grouping). Editable
    /// and independent of CreatedAt/UpdatedAt. Unix seconds; the admin renders it
    /// as a date picker (EditableDate).
    ArticleDate: EditableDate
    /// URL slug. Unique when present (Gen emits the unique index); the submit
    /// handler's UNIQUE-violation catch is backed by that index on every site.
    Slug: Unique<string> option
    CreatedAt: CreateTimestamp
    UpdatedAt: UpdateTimestamp option
    ViewCount: int
    DeletedAt: SoftDelete option
}

[<Table "comments">]
type ItemComment = {
    Id: PrimaryKey<string>
    ItemId: ForeignKey<Item>
    /// The shared app-level identity — decoupled from any concrete Identity type.
    IdentityId: IdentityRef
    /// The parent comment this reply targets (None for a top-level comment) — a
    /// typed self-reference, not a bare id string (wrap, don't unwrap).
    ParentId: ForeignKey<ItemComment> option
    Author: string
    Content: RichContent
    Removed: bool
    CreatedAt: CreateTimestamp
    DeletedAt: SoftDelete option
}

type Tag = {
    Id: PrimaryKey<string>
    Name: Unique<string>
    CreatedAt: CreateTimestamp
    DeletedAt: SoftDelete option
}

[<Table "item_tags">]
type ItemTag = {
    Id: PrimaryKey<string>
    ItemId: ForeignKey<Item>
    TagId: ForeignKey<Tag>
    DeletedAt: SoftDelete option
}

/// A captured snapshot of an item's source page — the microblog as its own web archive.
/// The captured bytes live in R2 (BlobKey); this row is the pointer + status. A side table
/// (FK -> Item), so the populated blog_items stays untouched (no in-place-FK rebuild).
[<Table "snapshots">]
type ItemSnapshot = {
    Id: PrimaryKey<string>
    ItemId: ForeignKey<Item>
    /// "html" (cleaned rendered DOM) for now; "screenshot" is a later variant.
    Kind: string
    /// R2 object key, e.g. "archive/<id>.html".
    BlobKey: string
    /// The source URL captured (denormalized from Item.Link at capture time).
    SourceUrl: string
    /// "pending" | "ok" | "failed".
    Status: string
    /// Failure detail for the admin, when Status = "failed".
    Error: string option
    /// Capture time (the row's creation). Named CreatedAt to match the framework's
    /// CreateTimestamp convention (column `created_at` + its auto-index).
    CreatedAt: CreateTimestamp
    DeletedAt: SoftDelete option
}

File: apps/microblog/extension/fable_output/packages/modules/blog/src/Models/Domain.js
import { Record } from "../../../../../fable_modules/fable-library-js.4.29.0/Types.js";
import { bool_type, record_type, int32_type, option_type, string_type } from "../../../../../fable_modules/fable-library-js.4.29.0/Reflection.js";
import { IdentityRef_$reflection, ForeignKey$1_$reflection, SoftDelete_$reflection, UpdateTimestamp_$reflection, CreateTimestamp_$reflection, Unique$1_$reflection, EditableDate_$reflection, RichContent_$reflection, Image_$reflection, Link_$reflection, PrimaryKey$1_$reflection } from "../../../../hedge/src/Hedge/Interface.js";

export class Item extends Record {
    constructor(Id, Title, Link, Image, Extract, OwnerComment, ArticleDate, Slug, CreatedAt, UpdatedAt, ViewCount, DeletedAt) {
        super();
        this.Id = Id;
        this.Title = Title;
        this.Link = Link;
        this.Image = Image;
        this.Extract = Extract;
        this.OwnerComment = OwnerComment;
        this.ArticleDate = ArticleDate;
        this.Slug = Slug;
        this.CreatedAt = CreatedAt;
        this.UpdatedAt = UpdatedAt;
        this.ViewCount = (ViewCount | 0);
        this.DeletedAt = DeletedAt;
    }
}

export function Item_$reflection() {
    return record_type("Blog.Domain.Item", [], Item, () => [["Id", PrimaryKey$1_$reflection(string_type)], ["Title", string_type], ["Link", option_type(Link_$reflection())], ["Image", option_type(Image_$reflection())], ["Extract", option_type(RichContent_$reflection())], ["OwnerComment", RichContent_$reflection()], ["ArticleDate", EditableDate_$reflection()], ["Slug", option_type(Unique$1_$reflection(string_type))], ["CreatedAt", CreateTimestamp_$reflection()], ["UpdatedAt", option_type(UpdateTimestamp_$reflection())], ["ViewCount", int32_type], ["DeletedAt", option_type(SoftDelete_$reflection())]]);
}

export class ItemComment extends Record {
    constructor(Id, ItemId, IdentityId, ParentId, Author, Content, Removed, CreatedAt, DeletedAt) {
        super();
        this.Id = Id;
        this.ItemId = ItemId;
        this.IdentityId = IdentityId;
        this.ParentId = ParentId;
        this.Author = Author;
        this.Content = Content;
        this.Removed = Removed;
        this.CreatedAt = CreatedAt;
        this.DeletedAt = DeletedAt;
    }
}

export function ItemComment_$reflection() {
    return record_type("Blog.Domain.ItemComment", [], ItemComment, () => [["Id", PrimaryKey$1_$reflection(string_type)], ["ItemId", ForeignKey$1_$reflection(Item_$reflection())], ["IdentityId", IdentityRef_$reflection()], ["ParentId", option_type(ForeignKey$1_$reflection(ItemComment_$reflection()))], ["Author", string_type], ["Content", RichContent_$reflection()], ["Removed", bool_type], ["CreatedAt", CreateTimestamp_$reflection()], ["DeletedAt", option_type(SoftDelete_$reflection())]]);
}

export class Tag extends Record {
    constructor(Id, Name, CreatedAt, DeletedAt) {
        super();
        this.Id = Id;
        this.Name = Name;
        this.CreatedAt = CreatedAt;
        this.DeletedAt = DeletedAt;
    }
}

export function Tag_$reflection() {
    return record_type("Blog.Domain.Tag", [], Tag, () => [["Id", PrimaryKey$1_$reflection(string_type)], ["Name", Unique$1_$reflection(string_type)], ["CreatedAt", CreateTimestamp_$reflection()], ["DeletedAt", option_type(SoftDelete_$reflection())]]);
}

export class ItemTag extends Record {
    constructor(Id, ItemId, TagId, DeletedAt) {
        super();
        this.Id = Id;
        this.ItemId = ItemId;
        this.TagId = TagId;
        this.DeletedAt = DeletedAt;
    }
}

export function ItemTag_$reflection() {
    return record_type("Blog.Domain.ItemTag", [], ItemTag, () => [["Id", PrimaryKey$1_$reflection(string_type)], ["ItemId", ForeignKey$1_$reflection(Item_$reflection())], ["TagId", ForeignKey$1_$reflection(Tag_$reflection())], ["DeletedAt", option_type(SoftDelete_$reflection())]]);
}

export class ItemSnapshot extends Record {
    constructor(Id, ItemId, Kind, BlobKey, SourceUrl, Status, Error$, CreatedAt, DeletedAt) {
        super();
        this.Id = Id;
        this.ItemId = ItemId;
        this.Kind = Kind;
        this.BlobKey = BlobKey;
        this.SourceUrl = SourceUrl;
        this.Status = Status;
        this.Error = Error$;
        this.CreatedAt = CreatedAt;
        this.DeletedAt = DeletedAt;
    }
}

export function ItemSnapshot_$reflection() {
    return record_type("Blog.Domain.ItemSnapshot", [], ItemSnapshot, () => [["Id", PrimaryKey$1_$reflection(string_type)], ["ItemId", ForeignKey$1_$reflection(Item_$reflection())], ["Kind", string_type], ["BlobKey", string_type], ["SourceUrl", string_type], ["Status", string_type], ["Error", option_type(string_type)], ["CreatedAt", CreateTimestamp_$reflection()], ["DeletedAt", option_type(SoftDelete_$reflection())]]);
}

